pkg/virtctl/vmexport/vmexport.go

```diff
@@ -0,0 +1,628 @@
+	roots := x509.NewCertPool()
+	roots.AppendCertsFromPEM([]byte(cert))
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{RootCAs: roots},
+	}
+	httpClient := httpClientCreatorFunc(transport)
+
+	// Generate and do the request
+	req, _ := http.NewRequest("GET", downloadUrl, nil)
+	req.Header.Set(exportTokenHeader, token)
+
+	return httpClient.Do(req)
+}
+
+// getHTTPClient assigns the default, non-mocked HTTP client
+func getHTTPClient(transport *http.Transport) *http.Client {
+	if insecure == true {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+
+	client := &http.Client{Transport: transport}
+	return client
+}
+
+// copyFileWithProgressBar serves as a wrapper to copy the file with a progress bar
+func copyFileWithProgressBar(output *os.File, resp *http.Response) error {
+	barTemplate := fmt.Sprintf(`{{ "Downloading file:" }} {{counters . }} {{ cycle . %s }} {{speed . }}`, progressBarCycle)
+
+	// start bar based on our template
+	bar := pb.ProgressBarTemplate(barTemplate).Start(0)
+	defer bar.Finish()
+	rd := bar.NewProxyReader(resp.Body)
+	bar.Start()
+
+	_, err := io.Copy(output, rd)
+	return err
+}
+
+// getOrCreateTokenSecret obtains a token secret to be used along with the virtualMachineExport
+func getOrCreateTokenSecret(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMachineExport) (*k8sv1.Secret, error) {
+	// Generate a random, 20 char string to be used as a token
+	token := rand.String(20)
+
+	secret := &k8sv1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getExportSecretName(),
+			Namespace: vmexport.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(vmexport, schema.GroupVersionKind{
+					Group:   exportv1.SchemeGroupVersion.Group,
+					Version: exportv1.SchemeGroupVersion.Version,
+					Kind:    "VirtualMachineExport",
+				}),
+			},
+		},
+		Type: k8sv1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			secretTokenKey: []byte(token),
+		},
+	}
+
+	_, err := client.CoreV1().Secrets(vmexport.Namespace).Create(context.Background(), secret, metav1.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// getTokenFromSecret extracts the token from the secret specified on the virtualMachineExport
+func getTokenFromSecret(client kubecli.KubevirtClient, namespace string) (string, error) {
+	secretName := getExportSecretName()
+	secret, err := client.CoreV1().Secrets(namespace).Get(context.Background(), secretName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	token := secret.Data[secretTokenKey]
+	return string(token), nil
+}
+
+// handleCreateFlags ensures that only compatible flag combinations are used with 'create' or with 'download' with the --create flag
+func handleCreateFlags(args []string) error {
+	if vm == "" && snapshot == "" && pvc == "" {
+		return fmt.Errorf(ErrRequiredExportType)
+	}
+
+	if vm != "" {
```

Comment:
Might also want to check these flags are mutually exclusive. Aka you cannot have `--vm=xxx` and `--pvc=yyy` The way I read it right now, if I do both, pvc will win.

Reply:
Yeah, I already handled that with cobra here: https://github.com/kubevirt/kubevirt/blob/a84658d561390aea9dab8090025835e1d6ce99df/pkg/virtctl/vmexport/vmexport.go#L175